File: asciivison/asciivision-wasm/src/processor.rs
use crate::ascii::gray_to_ascii;
use crate::config::Config;

/// Apply brightness and contrast adjustment to a grayscale value
#[inline]
fn apply_brightness_contrast(gray: u8, brightness: f32, contrast: f32) -> u8 {
    // Apply contrast around midpoint (127.5)
    let mut adjusted = (gray as f32 - 127.5) * contrast + 127.5;

    // Apply brightness (-1.0 to 1.0 maps to -255 to +255)
    adjusted += brightness * 255.0;

    // Clamp to valid range
    adjusted.clamp(0.0, 255.0) as u8
}

/// Convert RGBA pixel to grayscale using luminance formula
/// Same as original: 0.299*R + 0.587*G + 0.114*B
#[inline]
fn rgba_to_gray(r: u8, g: u8, b: u8) -> u8 {
    ((r as f32 * 0.299) + (g as f32 * 0.587) + (b as f32 * 0.114)) as u8
}

/// Process a frame of RGBA pixels and produce ASCII output
pub fn process_frame(
    config: &Config,
    pixels: &[u8],
    src_width: u32,
    src_height: u32,
    out_width: u32,
    out_height: u32,
    output: &mut String,
) {
    output.clear();

    // Calculate scaling factors
    // ASCII chars are ~2x taller than wide, so we sample more Y pixels
    let scale_x = src_width as f32 / out_width as f32;
    let scale_y = src_height as f32 / out_height as f32;

    let bytes_per_row = src_width as usize * 4; // RGBA = 4 bytes

    for y in 0..out_height {
        for x in 0..out_width {
            // Mirror horizontally for natural webcam feel
            let src_x = ((out_width - 1 - x) as f32 * scale_x) as usize;
            let src_y = (y as f32 * scale_y) as usize;

            // Bounds checking
            let src_x = src_x.min(src_width as usize - 1);
            let src_y = src_y.min(src_height as usize - 1);

            // RGBA format: R at offset 0, G at 1, B at 2, A at 3
            let pixel_offset = src_y * bytes_per_row + src_x * 4;

            if pixel_offset + 2 >= pixels.len() {
                output.push(' ');
                continue;
            }

            let r = pixels[pixel_offset];
            let g = pixels[pixel_offset + 1];
            let b = pixels[pixel_offset + 2];

            // Convert to grayscale
            let mut gray = rgba_to_gray(r, g, b);

            // Apply brightness/contrast
            gray = apply_brightness_contrast(gray, config.brightness, config.contrast);

            // Apply inversion if enabled
            if config.invert {
                gray = 255 - gray;
            }

            // Map to ASCII character
            let ascii_char = gray_to_ascii(gray, config.use_detailed_ramp);
            output.push(ascii_char);
        }
        output.push('\n');
    }
}

File: asciivision/asciivision-wasm/c/processor.c
#include "processor.h"
#include "ascii.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#define EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define EXPORT
#endif

// Clamp float to range
static inline float clampf(float v, float min, float max) {
    return v < min ? min : (v > max ? max : v);
}

// Convert RGBA to grayscale using luminance formula
static inline uint8_t rgba_to_gray(uint8_t r, uint8_t g, uint8_t b) {
    return (uint8_t)(r * 0.299f + g * 0.587f + b * 0.114f);
}

// Apply brightness and contrast adjustment
static inline uint8_t apply_brightness_contrast(uint8_t gray, float brightness, float contrast) {
    float adjusted = (gray - 127.5f) * contrast + 127.5f;
    adjusted += brightness * 255.0f;
    if (adjusted < 0.0f) return 0;
    if (adjusted > 255.0f) return 255;
    return (uint8_t)adjusted;
}

// Initialize config to defaults
static void config_default(Config* c) {
    c->brightness = 0.0f;
    c->contrast = 1.0f;
    c->use_detailed_ramp = 1;
    c->invert = 0;
}

EXPORT AsciiProcessor* processor_new(void) {
    AsciiProcessor* p = (AsciiProcessor*)malloc(sizeof(AsciiProcessor));
    if (!p) return NULL;

    config_default(&p->config);
    p->buffer_cap = 200 * 100 + 100;  // Same as Rust: 200*100 + 100 for newlines
    p->output_buffer = (char*)malloc(p->buffer_cap);
    p->status_buffer = (char*)malloc(128);

    if (!p->output_buffer || !p->status_buffer) {
        free(p->output_buffer);
        free(p->status_buffer);
        free(p);
        return NULL;
    }

    p->output_buffer[0] = '\0';
    p->status_buffer[0] = '\0';
    return p;
}

EXPORT void processor_free(AsciiProcessor* p) {
    if (p) {
        free(p->output_buffer);
        free(p->status_buffer);
        free(p);
    }
}

EXPORT const char* processor_process_frame(
    AsciiProcessor* p,
    uint8_t* pixels,
    uint32_t src_width,
    uint32_t src_height,
    uint32_t out_width,
    uint32_t out_height
) {
    if (!p || !pixels) return "";

    // Calculate scaling factors
    float scale_x = (float)src_width / (float)out_width;
    float scale_y = (float)src_height / (float)out_height;
    uint32_t bytes_per_row = src_width * 4;  // RGBA = 4 bytes

    char* out = p->output_buffer;
    uint32_t out_idx = 0;
    uint32_t max_idx = p->buffer_cap - 1;

    for (uint32_t y = 0; y < out_height && out_idx < max_idx; y++) {
        for (uint32_t x = 0; x < out_width && out_idx < max_idx; x++) {
            // Mirror horizontally for natural webcam feel
            uint32_t src_x = (uint32_t)((out_width - 1 - x) * scale_x);
            uint32_t src_y = (uint32_t)(y * scale_y);

            // Bounds check
            if (src_x >= src_width) src_x = src_width - 1;
            if (src_y >= src_height) src_y = src_height - 1;

            // RGBA format: R at offset 0, G at 1, B at 2
            uint32_t pixel_offset = src_y * bytes_per_row + src_x * 4;
            uint32_t pixel_max = src_width * src_height * 4;

            if (pixel_offset + 2 >= pixel_max) {
                out[out_idx++] = ' ';
                continue;
            }

            uint8_t r = pixels[pixel_offset];
            uint8_t g = pixels[pixel_offset + 1];
            uint8_t b = pixels[pixel_offset + 2];

            // Convert to grayscale
            uint8_t gray = rgba_to_gray(r, g, b);

            // Apply brightness/contrast
            gray = apply_brightness_contrast(gray, p->config.brightness, p->config.contrast);

            // Apply inversion if enabled
            if (p->config.invert) {
                gray = 255 - gray;
            }

            // Map to ASCII character
            out[out_idx++] = gray_to_ascii(gray, p->config.use_detailed_ramp);
        }
        if (out_idx < max_idx) {
            out[out_idx++] = '\n';
        }
    }
    out[out_idx] = '\0';

    return p->output_buffer;
}

EXPORT void processor_set_brightness(AsciiProcessor* p, float value) {
    if (p) p->config.brightness = clampf(value, -1.0f, 1.0f);
}

EXPORT float processor_get_brightness(AsciiProcessor* p) {
    return p ? p->config.brightness : 0.0f;
}

EXPORT void processor_set_contrast(AsciiProcessor* p, float value) {
    if (p) p->config.contrast = clampf(value, 0.1f, 3.0f);
}

EXPORT float processor_get_contrast(AsciiProcessor* p) {
    return p ? p->config.contrast : 1.0f;
}

EXPORT void processor_set_use_detailed_ramp(AsciiProcessor* p, int value) {
    if (p) p->config.use_detailed_ramp = value ? 1 : 0;
}

EXPORT void processor_toggle_ramp(AsciiProcessor* p) {
    if (p) p->config.use_detailed_ramp = !p->config.use_detailed_ramp;
}

EXPORT void processor_set_invert(AsciiProcessor* p, int value) {
    if (p) p->config.invert = value ? 1 : 0;
}

EXPORT void processor_toggle_invert(AsciiProcessor* p) {
    if (p) p->config.invert = !p->config.invert;
}

EXPORT void processor_reset(AsciiProcessor* p) {
    if (p) config_default(&p->config);
}

EXPORT const char* processor_get_status(AsciiProcessor* p, uint32_t width, uint32_t height) {
    if (!p) return "";
    snprintf(p->status_buffer, 128,
        "[%ux%u] B:%.1f C:%.1f | b/B:bright c/C:contrast r:ramp i:inv +/-:size 0:reset",
        width, height, p->config.brightness, p->config.contrast);
    return p->status_buffer;
}
